File: xophone/InstrumentViewController.m
//
//  GameViewController.m
//  xophone
//
//  Created by Dan Bar-Yaakov on 7/21/15.
//  Copyright (c) 2015 Dan Bar-Yaakov. All rights reserved.
//

#import "InstrumentViewController.h"
#import "InstrumentScene.h"
#import "SettingsViewController.h"

@implementation SKScene (Unarchive)

+ (instancetype)unarchiveFromFile:(NSString *)file {
    NSString *nodePath = [[NSBundle mainBundle] pathForResource:file ofType:@"sks"];
    NSData *data = [NSData dataWithContentsOfFile:nodePath
                                          options:NSDataReadingMappedIfSafe
                                            error:nil];
    NSKeyedUnarchiver *arch = [[NSKeyedUnarchiver alloc] initForReadingWithData:data];
    [arch setClass:self forClassName:@"SKScene"];
    SKScene *scene = [arch decodeObjectForKey:NSKeyedArchiveRootObjectKey];
    [arch finishDecoding];
    return scene;
}

@end

@implementation InstrumentViewController

- (void)viewDidLoad
{
    [super viewDidLoad];

    SKView * skView = (SKView *)self.view;
    skView.showsFPS = NO;
    skView.showsNodeCount = NO;
    skView.ignoresSiblingOrder = YES;
    self.view.multipleTouchEnabled = YES;
    InstrumentScene *scene = [InstrumentScene unarchiveFromFile:@"InstrumentScene"];
    
    scene.scaleMode = SKSceneScaleModeResizeFill;
    
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(showSettings) name:@"showSettings" object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(hideMenu) name:@"hideMenu" object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(showMenu) name:@"showMenu" object:nil];
    
    [[NSNotificationCenter defaultCenter] addObserver:self // put here the view controller which has to be notified
                                             selector:@selector(orientationChanged:)
                                                 name:@"UIDeviceOrientationDidChangeNotification"
                                               object:nil];
    
    [self initMenu];
    
    [skView presentScene:scene];
}

- (void)orientationChanged:(NSNotification *)notification{
    settingsButton.frame = CGRectMake(self.view.bounds.size.width - 138, self.view.bounds.size.height - 138, 128, 128);
}

-(void)initMenu {
    UIImage *starImage = [UIImage imageNamed:@"showSettings.png"];
    settingsButton = [[UIButton alloc] init];
    [settingsButton setBackgroundImage:starImage forState:UIControlStateNormal];
    settingsButton.frame = CGRectMake(self.view.bounds.size.width - 138, self.view.bounds.size.height - 138, 128, 128);
    [settingsButton addTarget:self action:@selector(showSettings) forControlEvents:UIControlEventTouchUpInside];
    [self.view addSubview:settingsButton];
    
}

-(void)showSettings {
    [self dismissViewControllerAnimated:YES completion:nil];
}

- (BOOL)shouldAutorotate
{
    return YES;
}
-(void)hideMenu {
    [UIView animateWithDuration:0.25 animations:^{settingsButton.alpha = 0.0;} completion:^(BOOL finished){if (finished) settingsButton.hidden = true;}];
}

-(void)showMenu {
    settingsButton.frame = CGRectMake(self.view.bounds.size.width - 138, self.view.bounds.size.height - 138, 128, 128);
    settingsButton.hidden = false;
    [UIView animateWithDuration:0.25 animations:^{settingsButton.alpha = 1.0;}];
}

- (NSUInteger)supportedInterfaceOrientations
{
    if ([[UIDevice currentDevice] userInterfaceIdiom] == UIUserInterfaceIdiomPhone) {
        return UIInterfaceOrientationMaskAllButUpsideDown;
    } else {
        return UIInterfaceOrientationMaskAll;
    }
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
}

- (BOOL)prefersStatusBarHidden {
    return YES;
}

@end

File: xophone/InstrumentViewController.swift
//
//  InstrumentViewController.swift
//  monoleap
//
//  Created by Omer Elimelech on 16/09/2021.
//  Copyright © 2021 Dan Bar-Yaakov. All rights reserved.
//

import Foundation
import UIKit
import SpriteKit

@objc class InstrumentViewController: UIViewController {
    
    var settingsButton: UIButton = {
        let button = UIButton(type: .custom)
        button.setImage(UIImage(named: "settings_new"), for: .normal)
        return button
    }()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        setupObservers()
        let skView: SKView = SKView(frame: self.view.frame)
        skView.showsFPS = false
        skView.showsNodeCount = false
        skView.ignoresSiblingOrder = true
        self.view.isMultipleTouchEnabled = true
        skView.isMultipleTouchEnabled = true
        view.addSubview(skView)
        let scene = InstrumentScene()
        scene.anchorPoint = CGPoint(x: 0, y: 0)
        scene.scaleMode = .resizeFill
        skView.presentScene(scene)
        initMenuButton()
    }
    
    override var prefersStatusBarHidden: Bool {
        return true
    }
    
    func setupObservers() {
        NotificationCenter.default.addObserver(self, selector: #selector(showSettings), name: .showSettings, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(hideMenu), name: .hideMenu, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(showMenu), name: .showMenu, object: nil)
        
        NotificationCenter.default.addObserver(self, selector: #selector(orientationChanged(_:)), name: UIDevice.orientationDidChangeNotification, object: nil)
    }
    
    func initMenuButton() {
        settingsButton.addTarget(self, action: #selector(showSettings), for: .touchUpInside)
        updateSettingsButton()
        self.view.addSubview(settingsButton)
    }
    
    @objc func orientationChanged(_ notification: Notification) {
        updateSettingsButton()
    }
    
    func updateSettingsButton() {
        settingsButton.frame = CGRect(x: view.frame.width - 170, y: view.frame.height - 170, width: 140, height: 142)
    }
    
    @objc func showSettings() {
        self.dismiss(animated: true, completion: nil)
    }
    
    @objc func hideMenu() {
        UIView.animate(withDuration: 0.25) {
            self.settingsButton.alpha = 0
        } completion: { finished in
            self.settingsButton.isHidden = true
        }
    }
    
    @objc func showMenu() {
        settingsButton.isHidden = false
        UIView.animate(withDuration: 0.25) {
            self.settingsButton.alpha = 1
        }
    }
    
    
}

extension Notification.Name {
    static let showSettings = Notification.Name("showSettings")
    static let hideMenu = Notification.Name("hideMenu")
    static let showMenu = Notification.Name("showMenu")
}
